File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-3.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l = len(nums)
        small = 0
        large = l - 1

        while True:
            index = small + ((large - small) // 2)

            if nums[index] == target:
                return index
            elif nums[index] < nums[large]:
                if target > nums[large]:
                    large = index - 1
                elif target < nums[large]:
                    if target < nums[index]:
                        large = index - 1
                    else:
                        small = index + 1
                else:
                    return large
            else:
                if target < nums[small]:
                    small = index + 1
                elif target > nums[small]:
                    if target < nums[index]:
                        large = index - 1
                    else:
                        small = index + 1
                else:
                    return small

            if small > large:
                return -1

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-3.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int totalSum = 0;
        for (int n : nums) totalSum += n;

        // If sum is odd, we can't partition into two equal integers
        if (totalSum % 2 != 0) return false;
        int target = totalSum / 2;

        // bitset<MAX_POSSIBLE_SUM>
        // bits[i] will be 1 if sum 'i' is reachable
        std::bitset<10001> bits(1); // bit 0 is 1 (sum 0 is always reachable)

        for (int n : nums) {
            // This single line replaces your entire DFS.
            // It shifts all reachable sums by 'n' and ORs them with previous sums.
            bits |= (bits << n);
        }

        return bits[target];
    }
};
